File: swift/Tests/Bandwidth/BandwidthWireTests.swift
// SPDX-License-Identifier: MIT

import Foundation
import XCTest
@testable import AetherNetProtocol

/// Unit tests for the ABMF WIRE bindings: BandwidthProbe(53), BandwidthAck(54),
/// BandwidthGossip(55). Binary little-endian byte-identity gates + send/handle behaviour.
///
/// Mirrors the C# `BandwidthWireTests` case-for-case:
///   1. Byte-identity — each codec must serialize to EXACTLY the LITTLE-ENDIAN bytes pinned in
///      `fixtures/bandwidth/vectors.json` (bytes → lowercase hex). This is the cross-language gate.
///      Belt-and-braces inline cases pin the same bytes even if the fixture file moves.
///   2. Behaviour — directed probe/ack sends captured by the shared ``FakeMeshSender``; gossip
///      broadcast; handle() raises the matching callback (gossip stamps peerUhid from the source);
///      a wrong packet type is rejected.
///
/// ``BandwidthWireService`` is a `public actor`, so every drive method is `async`. Actor-isolated
/// getters (the callback boxes below are read on the main test task) are read into a `let` before
/// any `XCTUnwrap`/`XCTAssertEqual` — never `await` inside an autoclosure.
final class BandwidthWireTests: XCTestCase {

    // MARK: - Fakes / capture

    // Directed/broadcast sends are captured by the shared `FakeMeshSender`
    // (Tests/FakeMeshSender.swift), a `final class` whose `unicasts()` / `broadcasts()` return the
    // recorded packets. A per-test nested `actor` fake is deliberately avoided: an `actor`
    // conforming to the nonisolated `MeshSender` requirements trips Swift 6 `#ConformanceIsolation`.

    /// Thread-safe one-shot capture box for a `@Sendable` callback value.
    private final class Box<T: Sendable>: @unchecked Sendable {
        private let lock = NSLock()
        private var value: T?
        func set(_ v: T) { lock.lock(); value = v; lock.unlock() }
        func get() -> T? { lock.lock(); defer { lock.unlock() }; return value }
    }

    // MARK: - Byte-identity corpus

    private struct Vector: Decodable {
        let name: String
        let kind: String
        let sequence: UInt32?
        let sender_send_us: Int64?
        let receiver_receive_us: Int64?
        let receiver_send_us: Int64?
        let probe_bytes: Int32?
        let btlbw_bps: Int64?
        let rtprop_us: Int32?
        let confidence: UInt8?
        let expected_hex: String
    }

    private struct Corpus: Decodable {
        let description: String
        let vectors: [Vector]
    }

    /// Locate `fixtures/bandwidth/vectors.json` by walking up from this source file's directory
    /// (`#file`) to the repo root — independent of CWD, the same parent-traversal idiom the URI,
    /// Bandwidth, and VideoCallControl fixture drivers use.
    private func loadCorpus() throws -> Corpus {
        var url = URL(fileURLWithPath: #file).deletingLastPathComponent()
        for _ in 0..<10 {
            let candidate = url
                .appendingPathComponent("fixtures")
                .appendingPathComponent("bandwidth")
                .appendingPathComponent("vectors.json")
            if FileManager.default.fileExists(atPath: candidate.path) {
                let data = try Data(contentsOf: candidate)
                return try JSONDecoder().decode(Corpus.self, from: data)
            }
            let parent = url.deletingLastPathComponent()
            if parent.path == url.path { break }
            url = parent
        }
        XCTFail("Could not locate fixtures/bandwidth/vectors.json walking up from \(#file)")
        throw CocoaError(.fileNoSuchFile)
    }

    private static func hex(_ d: Data) -> String {
        d.map { String(format: "%02x", $0) }.joined()
    }

    // MARK: - Byte-identity gates (fixture-driven)

    /// Every fixture vector must serialize to EXACTLY its `expected_hex` bytes.
    func testWire_MatchesCanonicalVectors() throws {
        let corpus = try loadCorpus()
        XCTAssertFalse(corpus.vectors.isEmpty, "corpus has no vectors")

        for v in corpus.vectors {
            let bytes: Data
            switch v.kind {
            case "probe":
                bytes = BandwidthWireCodec.serializeProbe(BandwidthProbe(
                    sequence: try XCTUnwrap(v.sequence, "[\(v.name)] missing sequence"),
                    senderSendUs: try XCTUnwrap(v.sender_send_us, "[\(v.name)] missing sender_send_us")
                ))
            case "ack":
                bytes = BandwidthWireCodec.serializeAck(BandwidthProbeAck(
                    sequence: try XCTUnwrap(v.sequence, "[\(v.name)] missing sequence"),
                    senderSendUs: try XCTUnwrap(v.sender_send_us, "[\(v.name)] missing sender_send_us"),
                    receiverReceiveUs: try XCTUnwrap(v.receiver_receive_us, "[\(v.name)] missing receiver_receive_us"),
                    receiverSendUs: try XCTUnwrap(v.receiver_send_us, "[\(v.name)] missing receiver_send_us"),
                    senderReceiveUs: 999, // local-only — must NOT change the wire bytes
                    probeBytes: try XCTUnwrap(v.probe_bytes, "[\(v.name)] missing probe_bytes")
                ))
            case "gossip":
                let conf = try XCTUnwrap(BandwidthConfidence(rawValue: try XCTUnwrap(v.confidence, "[\(v.name)] missing confidence")),
                                         "[\(v.name)] bad confidence")
                bytes = BandwidthWireCodec.serializeGossip(BandwidthGossipPayload(
                    peerUhid: "peer",        // not on the wire
                    transportName: "tp",     // not on the wire
                    btlBwBps: try XCTUnwrap(v.btlbw_bps, "[\(v.name)] missing btlbw_bps"),
                    rtPropUs: Int64(try XCTUnwrap(v.rtprop_us, "[\(v.name)] missing rtprop_us")),
                    confidence: conf,
                    measuredAt: Date() // not on the wire
                ))
            default:
                XCTFail("[\(v.name)] unknown kind \(v.kind)")
                continue
            }
            XCTAssertEqual(Self.hex(bytes), v.expected_hex, "[\(v.name)] wire byte mismatch")
        }
    }

    // MARK: - Byte-identity gates (inline, mirroring the C# InlineData / literal cases)

    func testProbe_SerializesToCanonicalBytes() {
        XCTAssertEqual(
            Self.hex(BandwidthWireCodec.serializeProbe(BandwidthProbe(sequence: 42, senderSendUs: 1_700_000_000_000_000))),
            "2a00000000401e18240a0600"
        )
    }

    func testAck_SerializesToCanonicalBytes() {
        // senderReceiveUs (999) is local-only and must NOT change the wire bytes.
        let ack = BandwidthProbeAck(
            sequence: 42,
            senderSendUs: 1_700_000_000_000_000,
            receiverReceiveUs: 1_700_000_000_012_345,
            receiverSendUs: 1_700_000_000_013_000,
            senderReceiveUs: 999,
            probeBytes: 1200
        )
        XCTAssertEqual(
            Self.hex(BandwidthWireCodec.serializeAck(ack)),
            "2a00000000401e18240a060039701e18240a0600c8721e18240a0600b0040000"
        )
    }

    func testGossip_SerializesToCanonicalBytes() {
        // peerUhid/transportName/measuredAt are not on the wire.
        let g = BandwidthGossipPayload(
            peerUhid: "peer", transportName: "tp",
            btlBwBps: 5_000_000, rtPropUs: 25_000,
            confidence: .medium, measuredAt: Date(timeIntervalSince1970: 0)
        )
        XCTAssertEqual(Self.hex(BandwidthWireCodec.serializeGossip(g)), "404b4c0000000000a861000002")
    }

    func testAck_RoundTrips_SenderReceiveUsZeroed() throws {
        let back = try XCTUnwrap(BandwidthWireCodec.deserializeAck(
            BandwidthWireCodec.serializeAck(BandwidthProbeAck(
                sequence: 7, senderSendUs: 100, receiverReceiveUs: 200,
                receiverSendUs: 300, senderReceiveUs: 400, probeBytes: 512
            ))
        ))
        XCTAssertEqual(back.sequence, 7)
        XCTAssertEqual(back.senderSendUs, 100)
        XCTAssertEqual(back.receiverReceiveUs, 200)
        XCTAssertEqual(back.receiverSendUs, 300)
        XCTAssertEqual(back.senderReceiveUs, 0) // not on wire
        XCTAssertEqual(back.probeBytes, 512)
    }

    // MARK: - Behaviour

    /// sendProbe directed-sends a BandwidthProbe packet to the peer.
    func testSendProbe_EmitsDirectedProbe() async throws {
        let sender = FakeMeshSender(localUhid: "aether:a:01")
        let svc = BandwidthWireService(sender: sender)

        let ok = await svc.sendProbe("aether:b:02", probe: BandwidthProbe(sequence: 42, senderSendUs: 1_700_000_000_000_000))
        XCTAssertTrue(ok)

        let sends = sender.unicasts()
        XCTAssertEqual(sends.count, 1)
        let sent = try XCTUnwrap(sends.first)
        XCTAssertEqual(sent.packet.type, .bandwidthProbe)
        XCTAssertEqual(sent.nextHopUhid, "aether:b:02")
        XCTAssertEqual(sent.packet.destinationUhid, "aether:b:02")
        XCTAssertEqual(sent.packet.sourceUhid, "aether:a:01")
    }

    /// sendAck directed-sends a BandwidthAck packet to the prober.
    func testSendAck_EmitsDirectedAck() async throws {
        let sender = FakeMeshSender(localUhid: "aether:local:01")
        let svc = BandwidthWireService(sender: sender)

        let ack = BandwidthProbeAck(sequence: 1, senderSendUs: 2, receiverReceiveUs: 3,
                                    receiverSendUs: 4, senderReceiveUs: 5, probeBytes: 6)
        let ok = await svc.sendAck("aether:b:02", ack: ack)
        XCTAssertTrue(ok)

        let sends = sender.unicasts()
        XCTAssertEqual(sends.count, 1)
        XCTAssertEqual(try XCTUnwrap(sends.first).packet.type, .bandwidthAck)
    }

    /// broadcastGossip emits a gossip broadcast; handle() on it raises onGossipReceived with the
    /// peerUhid stamped from the packet source.
    func testBroadcastGossip_EmitsGossip_AndHandleRaisesEvent_WithSourcePeer() async throws {
        let sender = FakeMeshSender(localUhid: "aether:local:01")
        // Seed 3 peers so broadcast fan-out returns 3, matching the C# fake's fixed 3.
        sender.addPeer(PeerInfo(uhid: "aether:p:1"))
        sender.addPeer(PeerInfo(uhid: "aether:p:2"))
        sender.addPeer(PeerInfo(uhid: "aether:p:3"))
        let svc = BandwidthWireService(sender: sender)

        let g = BandwidthGossipPayload(peerUhid: "", transportName: "",
                                       btlBwBps: 5_000_000, rtPropUs: 25_000,
                                       confidence: .medium, measuredAt: Date(timeIntervalSince1970: 0))
        let reached = await svc.broadcastGossip(g)
        XCTAssertEqual(reached, 3)

        let casts = sender.broadcasts()
        XCTAssertEqual(casts.count, 1)
        var sent = try XCTUnwrap(casts.first)
        XCTAssertEqual(sent.type, .bandwidthGossip)

        let box = Box<BandwidthGossipPayload>()
        await svc.setOnGossipReceived { box.set($0) }
        sent.sourceUhid = "aether:peer:09"
        let ok = await svc.handle(sent)
        XCTAssertTrue(ok)

        let got = try XCTUnwrap(box.get())
        XCTAssertEqual(got.btlBwBps, 5_000_000)
        XCTAssertEqual(got.rtPropUs, 25_000)
        XCTAssertEqual(got.confidence, .medium)
        XCTAssertEqual(got.peerUhid, "aether:peer:09")
    }

    /// handle() on a BandwidthProbe raises onProbeReceived carrying the packet source.
    func testHandle_Probe_RaisesProbeReceived_WithSource() async throws {
        let svc = BandwidthWireService(sender: FakeMeshSender(localUhid: "aether:local:01"))
        let box = Box<BandwidthProbeReceived>()
        await svc.setOnProbeReceived { box.set($0) }

        let pkt = MeshPacket(
            type: .bandwidthProbe,
            sourceUhid: "aether:x:01",
            payload: BandwidthWireCodec.serializeProbe(BandwidthProbe(sequence: 9, senderSendUs: 123))
        )
        let ok = await svc.handle(pkt)
        XCTAssertTrue(ok)

        let got = try XCTUnwrap(box.get())
        XCTAssertEqual(got.probe.sequence, 9)
        XCTAssertEqual(got.fromUhid, "aether:x:01")
    }

    /// handle() on a BandwidthAck raises onAckReceived.
    func testHandle_Ack_RaisesAckReceived() async throws {
        let svc = BandwidthWireService(sender: FakeMeshSender(localUhid: "aether:local:01"))
        let box = Box<BandwidthProbeAck>()
        await svc.setOnAckReceived { box.set($0) }

        let pkt = MeshPacket(
            type: .bandwidthAck,
            sourceUhid: "aether:x:01",
            payload: BandwidthWireCodec.serializeAck(BandwidthProbeAck(
                sequence: 3, senderSendUs: 10, receiverReceiveUs: 20,
                receiverSendUs: 30, senderReceiveUs: 0, probeBytes: 64
            ))
        )
        let ok = await svc.handle(pkt)
        XCTAssertTrue(ok)

        let got = try XCTUnwrap(box.get())
        XCTAssertEqual(got.sequence, 3)
        XCTAssertEqual(got.probeBytes, 64)
    }

    /// A packet whose type is not a bandwidth type is rejected (returns false).
    func testHandle_WrongType_ReturnsFalse() async {
        let svc = BandwidthWireService(sender: FakeMeshSender(localUhid: "aether:local:01"))
        let ok = await svc.handle(MeshPacket(type: .data, payload: Data()))
        XCTAssertFalse(ok)
    }

    /// A bandwidth-typed packet with a too-short body is rejected (returns false).
    func testHandle_ShortPayload_ReturnsFalse() async {
        let svc = BandwidthWireService(sender: FakeMeshSender(localUhid: "aether:local:01"))
        let ok = await svc.handle(MeshPacket(type: .bandwidthProbe, sourceUhid: "aether:x:01", payload: Data([0x01, 0x02])))
        XCTAssertFalse(ok)
    }
}

File: tests/AetherNet.Core.Tests/BandwidthWireTests.cs
// SPDX-License-Identifier: MIT

using AetherNet.Bandwidth;
using AetherNet.Models;
using AetherNet.Protocol;
using AetherNet.Routing;
using Microsoft.Extensions.Logging.Abstractions;
using Xunit;

namespace AetherNet.Core.Tests;

/// <summary>
/// Unit tests for the ABMF WIRE bindings: BandwidthProbe(53), BandwidthAck(54), BandwidthGossip(55).
/// Binary little-endian byte-identity gates + send/handle behaviour.
/// </summary>
public sealed class BandwidthWireTests
{
    private sealed class FakeMeshSender : IMeshSender
    {
        public string LocalUhid { get; set; } = "aether:local:01";
        public List<(MeshPacket Packet, string NextHop)> Sends { get; } = [];
        public List<MeshPacket> Broadcasts { get; } = [];

        public IReadOnlyList<PeerInfo> GetConnectedPeers() => Array.Empty<PeerInfo>();
        public Task<bool> SendAsync(MeshPacket packet, string nextHopUhid, CancellationToken ct = default)
        {
            Sends.Add((packet, nextHopUhid));
            return Task.FromResult(true);
        }
        public Task<int> BroadcastAsync(MeshPacket packet, CancellationToken ct = default)
        {
            Broadcasts.Add(packet);
            return Task.FromResult(3);
        }
    }

    private static string Hex(byte[] b) => Convert.ToHexString(b).ToLowerInvariant();

    // ── Byte-identity gates ─────────────────────────────────────────────────

    [Fact]
    public void Probe_SerializesToCanonicalBytes()
        => Assert.Equal("2a00000000401e18240a0600",
            Hex(BandwidthWireCodec.SerializeProbe(new BandwidthProbe(42, 1700000000000000L))));

    [Fact]
    public void Ack_SerializesToCanonicalBytes()
    {
        // SenderReceiveUs (999) is local-only and must NOT change the wire bytes.
        var ack = new BandwidthProbeAck(42, 1700000000000000L, 1700000000012345L, 1700000000013000L, SenderReceiveUs: 999L, ProbeBytes: 1200);
        Assert.Equal("2a00000000401e18240a060039701e18240a0600c8721e18240a0600b0040000",
            Hex(BandwidthWireCodec.SerializeAck(ack)));
    }

    [Fact]
    public void Gossip_SerializesToCanonicalBytes()
    {
        // PeerUhid/TransportName/MeasuredAt are not on the wire.
        var g = new BandwidthGossipPayload("peer", "tp", 5000000L, 25000L, BandwidthConfidence.Medium, default);
        Assert.Equal("404b4c0000000000a861000002", Hex(BandwidthWireCodec.SerializeGossip(g)));
    }

    [Fact]
    public void Ack_RoundTrips_SenderReceiveUsZeroed()
    {
        var back = BandwidthWireCodec.DeserializeAck(BandwidthWireCodec.SerializeAck(
            new BandwidthProbeAck(7, 100, 200, 300, 400, 512)));
        Assert.Equal(7u, back.Sequence);
        Assert.Equal(100, back.SenderSendUs);
        Assert.Equal(200, back.ReceiverReceiveUs);
        Assert.Equal(300, back.ReceiverSendUs);
        Assert.Equal(0, back.SenderReceiveUs); // not on wire
        Assert.Equal(512, back.ProbeBytes);
    }

    // ── Behaviour ───────────────────────────────────────────────────────────

    [Fact]
    public async Task SendProbe_EmitsDirectedProbe()
    {
        var s = new FakeMeshSender { LocalUhid = "aether:a:01" };
        var svc = new BandwidthWireService(s, NullLogger<BandwidthWireService>.Instance);
        Assert.True(await svc.SendProbeAsync("aether:b:02", new BandwidthProbe(42, 1700000000000000L)));
        var sent = Assert.Single(s.Sends);
        Assert.Equal(PacketType.BandwidthProbe, sent.Packet.Type);
        Assert.Equal("aether:b:02", sent.NextHop);
    }

    [Fact]
    public async Task SendAck_EmitsDirectedAck()
    {
        var s = new FakeMeshSender();
        var svc = new BandwidthWireService(s, NullLogger<BandwidthWireService>.Instance);
        var ack = new BandwidthProbeAck(1, 2, 3, 4, 5, 6);
        Assert.True(await svc.SendAckAsync("aether:b:02", ack));
        Assert.Equal(PacketType.BandwidthAck, Assert.Single(s.Sends).Packet.Type);
    }

    [Fact]
    public async Task BroadcastGossip_EmitsGossip_AndHandleRaisesEvent_WithSourcePeer()
    {
        var s = new FakeMeshSender();
        var svc = new BandwidthWireService(s, NullLogger<BandwidthWireService>.Instance);
        var g = new BandwidthGossipPayload("", "", 5000000L, 25000L, BandwidthConfidence.Medium, default);
        Assert.Equal(3, await svc.BroadcastGossipAsync(g));
        var sent = Assert.Single(s.Broadcasts);
        Assert.Equal(PacketType.BandwidthGossip, sent.Type);

        BandwidthGossipPayload? got = null;
        svc.GossipReceived += (_, e) => got = e;
        sent.SourceUhid = "aether:peer:09";
        Assert.True(await svc.HandleAsync(sent));
        Assert.NotNull(got);
        Assert.Equal(5000000L, got!.BtlBwBps);
        Assert.Equal(25000L, got.RtPropUs);
        Assert.Equal(BandwidthConfidence.Medium, got.Confidence);
        Assert.Equal("aether:peer:09", got.PeerUhid);
    }

    [Fact]
    public async Task Handle_Probe_RaisesProbeReceived_WithSource()
    {
        var svc = new BandwidthWireService(new FakeMeshSender(), NullLogger<BandwidthWireService>.Instance);
        BandwidthProbeReceived? got = null;
        svc.ProbeReceived += (_, e) => got = e;
        var pkt = new MeshPacket
        {
            Type = PacketType.BandwidthProbe,
            SourceUhid = "aether:x:01",
            Payload = BandwidthWireCodec.SerializeProbe(new BandwidthProbe(9, 123)),
        };
        Assert.True(await svc.HandleAsync(pkt));
        Assert.NotNull(got);
        Assert.Equal(9u, got!.Probe.Sequence);
        Assert.Equal("aether:x:01", got.FromUhid);
    }

    [Fact]
    public async Task Handle_Ack_RaisesAckReceived()
    {
        var svc = new BandwidthWireService(new FakeMeshSender(), NullLogger<BandwidthWireService>.Instance);
        BandwidthProbeAck? got = null;
        svc.AckReceived += (_, e) => got = e;
        var pkt = new MeshPacket
        {
            Type = PacketType.BandwidthAck,
            SourceUhid = "aether:x:01",
            Payload = BandwidthWireCodec.SerializeAck(new BandwidthProbeAck(3, 10, 20, 30, 0, 64)),
        };
        Assert.True(await svc.HandleAsync(pkt));
        Assert.NotNull(got);
        Assert.Equal(3u, got!.Sequence);
        Assert.Equal(64, got.ProbeBytes);
    }

    [Fact]
    public async Task Handle_WrongType_ReturnsFalse()
    {
        var svc = new BandwidthWireService(new FakeMeshSender(), NullLogger<BandwidthWireService>.Instance);
        Assert.False(await svc.HandleAsync(new MeshPacket { Type = PacketType.Data, Payload = Array.Empty<byte>() }));
    }
}
